### PR TITLE
Unlock laser after sunk

### PR DESCRIPTION
**Pull request by Ben-Geyer**

### CODE REVIEW

File: src/main/java/controllers/ApplicationController.java [hunk 1]
```diff
@@ -32,7 +32,15 @@ public Result placeShip(Context context, PlacementGameAction g) {
 
     public Result attack(Context context, AttackGameAction g) {
         Game game = g.getGame();
-        boolean result = game.attack(g.getActionRow(), g.getActionColumn(), g.getActionSonar());
+        boolean result;
```
Comment (mateoreyrosa) [inline src/main/java/controllers/ApplicationController.java:35]: Noted that you added direction to the front end's view model

File: src/main/java/cs361/battleships/models/Board.java [hunk 2]
```diff
@@ -114,6 +114,15 @@ public Result sonarPulse(int x, char y) {
 		return finalResult;
 	}
 
+	public Result moveShips(char moveDir) {
+		for (Ship s : ships) {
```
Comment (mateoreyrosa) [inline src/main/java/cs361/battleships/models/Board.java]: Good job breaking down the moveShips functionality into its own function instead of trying to attach it to another process

File: src/main/java/cs361/battleships/models/Square.java [hunk 3]
```diff
@@ -19,6 +19,19 @@ public Square(int row, char column) {
 		this.column = column;
 	}
 
+	public void move(char moveDir) {
+		switch(moveDir) {
```
Comment (mateoreyrosa) [inline src/main/java/cs361/battleships/models/Square.java:30]: good use of a switch statement for each of the directions. Im not sure that there is really a better way to store the individual directions except for maybe an array

File: src/test/java/cs361/battleships/models/BoardTest.java [hunk 4]
```diff
@@ -145,4 +145,43 @@ public void testSonarPulseRightAmountOfSquaresSides() {
         board.sonarPulse(5, 'J');
         assertEquals(board.getAttacks().size(), 36);
     }
+
+    @Test
+    public void testMoveCorrectDirection() {
+        board.placeShip(new Ship("MINESWEEPER"), 6, 'D', true);
```
Comment (mateoreyrosa) [inline src/test/java/cs361/battleships/models/BoardTest.java:152]: Good thorough tests for each direction, increases confidence of code functionality

File: target/classes/assets/css/custom.css [hunk 5]
```diff
@@ -1,8 +1,25 @@
 /* Add additional stylesheets below
 -------------------------------------------------- */
+
 .captain{
 background-color: brown !important;
 }
+
```
Comment (mateoreyrosa) [inline target/classes/assets/css/custom.css:7]: I think these changes were already in

File: src/main/java/cs361/battleships/models/Ship.java [hunk 6]
```diff
@@ -117,10 +114,21 @@ else if(attackedSquare.getIsCaptain()){
 
 		}
 	}
+
+	public boolean validateMove(char moveDir) {
+		int tRow = occupiedSquares.stream().max(Comparator.comparing(s -> s.getRow())).get().getRow();
```
Comment (mateoreyrosa) [inline src/main/java/cs361/battleships/models/Ship.java]: Looks like all of the bases are covered for this validation function. Good use of the stream on the occupiedSquares, simplifies the code a lot.

File: src/test/java/cs361/battleships/models/ShipTest.java [hunk 7]
```diff
@@ -209,4 +209,55 @@ public void testMoveOneSquareCorrectDirection() {
         assertEquals(s.getOccupiedSquares().get(1).getRow(), 6);
         assertEquals(s.getOccupiedSquares().get(1).getColumn(), 'E');
     }
+
+    @Test
+    public void testDoNotMoveOverEdge() {
```
Comment (mateoreyrosa) [inline src/test/java/cs361/battleships/models/ShipTest.java:214]: I like the thoroughness of the tests because many edge cases are covered.

File: src/main/java/cs361/battleships/models/Board.java [hunk 8]
```diff
@@ -2,8 +2,8 @@
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 
-import java.util.ArrayList;
-import java.util.List;
+import java.lang.reflect.Array;
+import java.util.*;
```
Comment (mateoreyrosa) [inline src/main/java/cs361/battleships/models/Board.java:6]: Is the .* adding all of the java utils? may not be necessary. (bloating)

File: src/main/java/cs361/battleships/models/Ship.java [hunk 9]
```diff
@@ -37,10 +37,24 @@ public Ship(String kind) {
 		}
 	}
 
+	public Ship(Ship other) {
```
Comment (mateoreyrosa) [inline src/main/java/cs361/battleships/models/Ship.java:40]: Good additions to the copy constructor.

File: src/test/java/cs361/battleships/models/ShipTest.java [hunk 10]
```diff
@@ -260,4 +260,27 @@ public void testDoNotMoveOverEdge() {
         assertEquals(s.getOccupiedSquares().get(1).getRow(), 2);
         assertEquals(s.getOccupiedSquares().get(1).getColumn(), 'J');
     }
+
+    @Test
+    public void testNoMoveOverlap() {
```
Comment (mateoreyrosa) [inline src/test/java/cs361/battleships/models/ShipTest.java:265]: Good checks, might need to modify for the submarine that submerges.

File: src/test/java/cs361/battleships/models/BoardTest.java [hunk 11]
```diff
@@ -227,4 +227,15 @@ public void testNoMoveOverlap() {
         assertEquals(board.getShips().stream().filter(s -> s.getKind() == "DESTROYER").findFirst().get().getOccupiedSquares().get(0).getRow(), 8);
         assertEquals(board.getShips().stream().filter(s -> s.getKind() == "DESTROYER").findFirst().get().getOccupiedSquares().get(0).getColumn(), 'H');
     }
+
+    @Test
+    public void testAttackWithSpaceLaser() {
```
Comment (mateoreyrosa) [inline src/test/java/cs361/battleships/models/BoardTest.java:232]: Good test

File: src/main/java/assets/css/custom.css [hunk 12]
```diff
@@ -176,6 +176,20 @@ text-align:center;
     border: 3px dashed rgba(178, 103, 253, 1);
 }
 
+#opponent.laser td:hover {
+    background-color: rgba(73, 250, 15, 0.4);
```
Comment (mateoreyrosa) [inline src/main/java/assets/css/custom.css:180]: clear css, we may want to clear some of the styles up since they are a little convoluted

File: src/main/java/assets/game.js [hunk 13]
```diff
@@ -26,15 +26,46 @@ function checkSunk(elementId, ship) {
             var cell = document.getElementById("opponent").rows[square.row-1].cells[square.column.charCodeAt(0) - 'A'.charCodeAt(0)];
             var div = cell.querySelector("div");
 
-            if (i == 0) {
+            if (i == 0 && div != null) {
+                div = document.createElement('div');
```
Comment (mateoreyrosa) [inline src/main/java/assets/game.js:30]: Good additions to the front end code, null checks make game run smoothly

**Review by mateoreyrosa — APPROVED** [review]
Overall the code is concise and leverages the existing classes to create concise backend code. All of the tests were very thorough and covered edge cases that could potentially break the program. There was not too much that was changed in terms of class structure, as it was not really necessary for this portion, just some added functions. Some improvements may include watching for necessary import statements.
